Code/GraphMol/SubstructLibrary/SubstructLibrary.h

```diff
@@ -0,0 +1,310 @@
+  }
+  const FPHolderBase & getFingerprints() const {
+    PRECONDITION(fps, "No fingerprints set in library");
+    return *fps;
+  }
+
+  //!Add a molecule to the library
+  /*!
+    \param mol Molecule to add
+  */
+  unsigned int addMol(const ROMol &mol);
+
+  //!Get the matching indices for the query
+  /*!
+    \param query       Query to match against molecules
+    \param numThreads  If -1 use all available processors [default -1]
+    \param maxResults  Maximum results to return, -1 means return all [default -1]    
+  */
+  std::vector<unsigned int> getMatches(const ROMol &query, int numThreads=-1,
+                                       int maxResults=-1);
+  //!Get the matching indices for the query between the given indices
+  /*!
+    \param query       Query to match against molecules
+    \param startIdx    Start index of the search
+    \param endIdx      Ending idx (inclusive) of the search.
+    \param numThreads  If -1 use all available processors [default -1]
+    \param maxResults  Maximum results to return, -1 means return all [default -1]
+  */  
+  std::vector<unsigned int> getMatches(const ROMol &query,
+                                       unsigned int startIdx, unsigned int endIdx,
+                                       int numThreads=-1,
+                                       int maxResults=-1);
+
+  //!Return the number of matches for the query
+  /*!
+    \param query       Query to match against molecules
+    \param numThreads  If -1 use all available processors [default -1]
+  */  
+  unsigned int countMatches(const ROMol &query, int numThreads=-1);
+  //!Return the number of matches for the query between the given indices
+  /*!
+    \param query       Query to match against molecules
+    \param startIdx    Start index of the search
+    \param endIdx      Ending idx (inclusive) of the search.
+    \param numThreads  If -1 use all available processors [default -1]
+  */    
+  unsigned int countMatches(const ROMol &query,
+                            unsigned int startIdx, unsigned int endIdx, int numThreads=-1);
+
+  //! Returns true if any match exists for the query
+  /*!
+    \param query       Query to match against molecules
+    \param numThreads  If -1 use all available processors [default -1]
+  */  
+  bool hasMatch(const ROMol &query, int numThreads=-1);
+  //! Returns true if any match exists for the query between the specified indices
+  /*!
+    \param query       Query to match against molecules
+    \param startIdx    Start index of the search
+    \param endIdx      Ending idx (inclusive) of the search.
+    \param numThreads  If -1 use all available processors [default -1]
+  */      
+  bool hasMatch(const ROMol &query,
+                unsigned int startIdx, unsigned int endIdx, int numThreads=-1);
+
+
+  //! Returns the molecule at the given index
+  /*!
+    \param idx       Index of the molecule in the library
+  */
+  boost::shared_ptr<ROMol> getMol(unsigned int idx) const {
```

Comment:
As mentioned mols can't be null.  I just do this so I can send the raw pointer to the thread, if this is confusing I can get rid of the variable and do it internally.  Might be clearer as well.